0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+invalid inputs in a pre-condition, dynamically fail if given invalid strings, or implement a lossy encoding by
+discarding or replacing invalid characters.
+
+## Compatibility
+
+The current definitions of `string` and `char` are biased towards using a UTF-16 encoding at runtime.
+This aligns well with some compilation target languages which also use UTF-16, such as Java, C#, and JavaScript, 
+but less well with those that use the recently more popular UTF-8 encoding, such as Go or Rust.
+Any Dafny code that interfaces with external code will often have to convert between `string` values and
+native representations of strings, and baking in the assumption of UTF-16 imposes a complexity and performance
+penalty for multiple target environments.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As of Dafny 4.0, the default semantics of the `char` type have been changed to represent Unicode scalar values instead of UTF-16 code points.
+This means that surrogate code points, meaning those in the range `[0xD800, 0xDFFF]`, are no longer allowed. This also means
+code points that require more than 16 bits, meaning those in the range `[0x10000, 0x1FFFF]`, are now directly representable
+as `char` values.
+
+The definition of the `char` type is controlled by a command-line switch called `/unicodeChar`, where `/unicodeChar:0`
+enables the previous semantics and `/unicodeChar:1` enables the new semantics described above.
+The default prior to Dafny 4.0 is `/unicodeChar:0`, but the default as of Dafny 4.0 is `/unicodeChar:1`.
+This change in definition can cause previously verified code to no longer verify,
+since expressions such as `someIntValue as char` or `someCharValue + 'a'` are no longer allowed to result in
+surrogate code points.
+
+This change can also result in behavioral changes that may not be caught by verification,
+as they may not intersect with any explicit specifications.
+Here is an example of a program that will behave differently depending on the value of `/unicodeChar`,
+and hence differently on Dafny 4.0 compared to earlier versions:
+
+```dafny
+method Main() {
+  var s := "Unicode is just so \ud83d\ude0e";
+  print s, "\n"     // "Unicode is just so 😎" (not affected by /unicodeChar)
+  print |s|, "\n"   // /unicodeChar:0 => "21"
+                    // /unicodeChar:1 => "20"
+  print s[19], "\n" // /unicodeChar:0 => "�"
+                    // /unicodeChar:1 => "😎"
+}
+```
+
+A related but largely orthogonal issue has also been addressed.
+Dafny source code files must be encoded in UTF-8,
+but in previous versions of Dafny,
+string and character literals could only contain printable and white-space ASCII characters,
+due to a limitation of the Coco/R parser generator the toolchain uses.
+This has been fixed, and both standard form and verbatim form string literals now allow any Unicode characters.
```

Comment:
What has been fixed? The Coco/R parser?

Reply:
This section is intended to describe the desired end state for Dafny users, so it's describing a future state. :)

I'm saying I can see where the issue is (specifically in our copy of [`Scanner.frame`](https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/Coco/Scanner.frame#L318), which is creating a `Buffer` when it should be creating a `UTF8Buffer`) so I'm proposing we fix that.

Reply:
FYI this part has been actually fixed now: https://github.com/dafny-lang/dafny/pull/2717 :)